File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.java
class Solution {
    public int numberOfSubstrings(String s) {
        HashMap<Character,Integer> mp=new HashMap<>();
        int start=0;
        int end=0;
        int ans=0;
        while(end<s.length()){
            mp.put(s.charAt(end),mp.getOrDefault(s.charAt(end),0)+1);
            while(mp.size()==3){
                ans+=s.length()-end;
                mp.put(s.charAt(start),mp.get(s.charAt(start))-1);
                if(mp.get(s.charAt(start))==0){
                    mp.remove(s.charAt(start));
                }
                start++;
            }
            end++;
        }
        return ans;
    }
}

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        return count(s);
    }
    int count(string s){
        int end=0;
        int start=0;
        int ans=0;
        vector<int> freq(3,0);
        int n=s.length();
        while(end<n){
            freq[s[end]-'a']++;
            while(freq[0]>0 && freq[1]>0 && freq[2]>0){
                ans+=n-end;
                freq[s[start]-'a']--;
                 start++;
            }
            end++;
        }
        return ans;
    }
};